pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,380 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name: certSecretName,
+			},
+			Data: map[string][]byte{
+				"tls.key": keyPEM,
+			},
+		}, metav1.CreateOptions{})
+
+		return key, nil, true, err
+	}
+
+	certPEM := certSecret.Data["tls.crt"]
+	if len(certPEM) > 0 {
+		cert, err = pemDecodeCert(certPEM)
+
+		// delete tls.crt if it's malformed or its pub key does not match the private key
+		var deleteCert bool
+		if err != nil {
+			klog.Error("failed to decode the certificate used for the openshift-authenticator, new certificate will be requested: %v", err)
+			deleteCert = true
+		} else if !key.PublicKey.Equal(cert.PublicKey.(*rsa.PublicKey)) {
+			klog.Error("the public key of the certificate used for the openshift-authenticator does not match the stored private key, new certificate will be requested: %v", err)
+			deleteCert = true
+		}
+
+		if deleteCert {
+			certSecretCopy := certSecret.DeepCopy()
+			delete(certSecretCopy.Data, "tls.crt")
+			_, err := c.secretClient.Update(ctx, certSecretCopy, metav1.UpdateOptions{})
+			return key, nil, false, err
+		}
+	}
+
+	return key, cert, false, nil
+}
+
+// resubmitCSR deletes the old CSR API object and creates a new one for the given key
+func (c *webhookAuthenticatorCertRequester) resubmitCSR(ctx context.Context, key *rsa.PrivateKey) error {
+	err := c.csrClient.Delete(ctx, csrName, metav1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	csrDER, err := x509.CreateCertificateRequest(rand.Reader,
+		&x509.CertificateRequest{
+			Subject: pkix.Name{CommonName: "system:serviceaccount:openshift-oauth-apiserver:openshift-authenticator"},
+		},
+		key,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create a CSR: %w", err)
+	}
+
+	csrPEM, err := pemEncodeCSR(csrDER)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.csrClient.Create(ctx,
+		&certapiv1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: csrName,
+			},
+			Spec: certapiv1.CertificateSigningRequestSpec{
+				Request:    csrPEM,
+				SignerName: certapiv1.KubeAPIServerClientSignerName,
+				Usages: []certapiv1.KeyUsage{
+					certapiv1.UsageDigitalSignature,
+					certapiv1.UsageKeyEncipherment,
+					certapiv1.UsageClientAuth,
+				},
+			},
+		},
+		metav1.CreateOptions{},
+	)
+
+	return err
+}
+
+func nameFilter(name string) factory.EventFilterFunc {
+	return func(obj interface{}) bool {
+		metaObj, ok := obj.(metav1.ObjectMetaAccessor)
+		if !ok {
+			klog.Error("it's not a metaobj: %#v", obj)
+			return false
+		}
+		return metaObj.GetObjectMeta().GetName() == name
+	}
+}
+
+func pemEncodeCSR(csrDER []byte) ([]byte, error) {
+	buffer := bytes.Buffer{}
+
+	if err := pem.Encode(&buffer,
+		&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrDER},
+	); err != nil {
+		return nil, fmt.Errorf("failed to convert DER CSR into a PEM-formatted block: %w", err)
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func pemEncodeRSAPrivate(key *rsa.PrivateKey) ([]byte, error) {
```

Comment:
we don't have any convenience functions in `library-go` for crypto?

Reply:
none that are exposed for private key PEM-encoding, none exist for CSRs and RSA key parsing.